example/ram/compile.lurk

```diff
@@ -0,0 +1,71 @@
+(defmacro or (a b)
+  `(if ,a t ,b))
+
+(define map
+  (lambda (f xs)
+    (if (eq nil xs)
+        nil
+        (cons (f (car xs)) (map f (cdr xs))))))
+
+(define cadr
+  (lambda (xs)
+    (car (cdr xs))))
+
+(define caddr
+  (lambda (xs)
+    (car (cdr (cdr xs)))))
+
+(define cadddr
+  (lambda (xs)
+    (car (cdr (cdr (cdr xs))))))
+
+(define let-bindings cadr)
+(define let-body caddr)
+(define lambda-params cadr)
+(define lambda-body caddr)
+
+(define lookup
+  (lambda (e env success failure)
+    (if (eq nil env)
+        (failure)
+        (if (eq nil (atom (car (car env))))
+            (lookup e (car env) success
+                    (lambda () (lookup e (cdr env) success failure)))
+            (if (eq e (car (car env)))
+                (success (cdr (car env)))
+                (lookup e (cdr env) success failure))))))
+
+(define member
+  (lambda (e xs)
+    (if (eq nil xs)
+        nil
+        (if (eq e (car xs))
+            xs
+            (member e (cdr xs))))))
+
+(define compile-closure
+  (lambda (e params ram env)
+    (let ((init-params (append `(if current-env current-ram eval begin atom car cdr emit quote macroexpand + - / * = eq cons) params))
+          (rec (lambda (e) (compile-closure e init-params ram env)))
+          (rec-params (lambda (e params) (compile-closure e (append init-params params) ram env))))
+      (if (atom e)
+          (if (member e params)
+              e
+              (lookup e env (lambda (v)
+                              ;;(if (functionp v) e v)
+                              v
+                              )
+                      (lambda () (lookup e ram (lambda (v) v) (lambda () e)))))
+          (let ((tag (car e)))
+            (if (or (eq 'let tag) (eq 'letrec tag))
+                ;; NOTE: this over approximates params in the right-hand sides
+                (let ((all-params (append (map (lambda (b) (car b)) (let-bindings e)) params)))
+                  `(,tag
+                    ,(map (lambda (b) `(,(car b) ,(rec-params (cadr b) all-params))) (let-bindings e))
```

Comment:
We need all-params because if we have:
```
(let ((x 1))
  (let ((f (compile (lambda (y) (let ((x 2)) x)))) (f 0))
```
we want this to return 2 not 1. We don't want to inline the x from the env. Anything that is a parameter should not be inlined.

Reply:
I figured out what was confusing me. My concern was that all-params was using bindings introduced later when compiling values of those introduced earlier. But I now realize that it is only used in the body of the let for which all-params does indeed represent all the relevant params. Thanks.